### ISSUE TITLE
New Feature: Predict Mode

### ISSUE THREAD

**Issue by veillette** [phetsims/calculus-grapher#100]
The control fits above the “width” slider in the control panel on all screens

![image](https://user-images.githubusercontent.com/7633344/201945468-334c6352-0b0e-4a90-966e-2bebdeefaa18.png)

When in predict mode, all of the curve controls turn gray, and the label inside the f(x) graph updates to say “Predict f(x)”, and a “Show f(x)” checkbox appears. If on the Derivative or Integral screens, the Tangent or Area Under Curve checkboxes need to be  invisible, and any associated controls not visible. 

If viewing both f(x) and the prediction curve, be sure the gray prediction curve is layered on top of the blue f(x) curve. The gray curve should be 1.75px. 

![image](https://user-images.githubusercontent.com/7633344/201946843-361eece4-834d-40eb-9e4d-21e728819447.png)

**Comment by veillette** [comment]
@amanda-phet:

I have a comment that you should check for correctness and a question about model behavior.

Here is my understanding of the predictMode.
In PredictMode, the colors of the CurveManipulationIcons as well as witdthDisplayCurve changes color to reflect the fact that are in PredictMode. However, the value of the width as well as the selected curveManipulation mode are shared between f(x) and and Predict. In other words, if you selected a width of 6 in predict and toggle back to f(x), it would have a width of 6 as well.
 
How about the yellow buttons: We have an eraser, undo and smooth buttons.
What should happen if you press a yellow button in Predict Mode?
I see three possibilities.
-it applies the action only to f(x) at all times 
-it applies the action only to the currently selected curve
-it applies the same action to both curves

**Comment by pixelzoom** [comment]
My 2 cents:

> Here is my understanding of the predictMode. ...

Yes, what you've described should be the behavior. I see no reason to complicate things by having the 2 modes each have their own "width" state for the manipulations.

> How about the yellow buttons: We have an eraser, an undo and a smooth button.
What should happen if you press a yellow button in Predict Mode?

In Predict mode, the yellow push buttons should apply only to the prediction curve. The original f(x) should be unaffected.

**Comment by veillette** [comment]
I also support  @pixelzoom recommendation for the buttons as it provides the most user friendly experience.
- The user should expect an action if pressing a button, so option (1) is sub optimal since the user may not have turned the visibility of f(x).
- Options (3) would be surprising, particularly if a user makes a mistake while drawing a curve and would like to erase it.

**Comment by pixelzoom** [comment]
Other aspects of this feature that has not been specified:

* When Predict mode is turned on and off, does it remember the previous prediction curve, or start with a new curve each time?  I think it needs to remember, in case the student/teacher wants to toggle on/off while trying to match a specific f(x).

* What is the PhET-iO instrumentation? Is the prediction curve stateful?

**Comment by amanda-phet** [comment]
> In PredictMode, the colors of the CurveManipulationIcons as well as witdthDisplayCurve changes color to reflect the fact that are in PredictMode. However, the value of the width as well as the selected curveManipulation mode are shared between f(x) and and Predict. In other words, if you selected a width of 6 in predict and toggle back to f(x), it would have a width of 6 as well.

Yes, that behavior makes sense to me. 

> How about the yellow buttons: We have an eraser, an undo and a smooth button.
> What should happen if you press a yellow button in Predict Mode?
> I see three possibilities.
> 
> (1) it applies the action only to f(x) at all times
> (2) it applies the action only to the currently selected curve
> (3)it applies the same action to both curves

Definitely option (2). It should apply to the currently selected/interactive curve. 

> Other aspects of this feature that has not been specified:
> 
> When Predict mode is toggled on/off, does it remember the previous prediction curve, or start with a new prediction curve each time? I think it needs to remember, in case the student/teacher wants to toggle on/off while trying to match a specific f(x).

I think it should remember for the exact example you provided. 

> What is the PhET-iO instrumentation for the prediction curve? Is it stateful? Is it read-only?

Let's discuss this because I am not sure exactly what you are asking. 

Stateful I think I understand- do we want the prediction curve part of the saved state? I think so, because if it wasn't stateful that would be a strange user experience to me. 

Read-only- I'm not sure I understand the full implications of this question.

**Comment by pixelzoom** [comment]
@amanda-phet answered all of our questions, expect the statefulness of the prediction curve.  @veillette Unless it's going to be additional work (which I doubt), let's plan on the prediction curve being stateful.  In the current implementation, I believe that's the case by default.

**Comment by veillette** [comment]
`GraphNode` creates a `CurveNode`, but since `OriginalGraphNode` extends `GraphNode` and `OriginalGraphNode` add a predictCurveNode, it would be more appropriate for `GraphNode` to add a CurveLayer where all the curveNodes are added. The CurveLayer visiblity would the tied to the status EyeToggleButton.  The visibility of the individuals curve in OriginalGraphNode will be related to the status of the `PredictModeEnabledProperty` and `AllOriginalCurveVisibleProperty`. The last property would track the visibility of f(x).   

![image](https://user-images.githubusercontent.com/7633344/204897938-589a0ca7-d38d-4c14-a3bd-0bd981a8a981.png)

**Comment by veillette** [comment]
`AllOriginalCurvesVisibleProperty` seems a particularly bad name but I cant think of a better one.  The 'show f(x)' checkbox is only present when the predictMode is enabled. The putative 'predictModeEnableProperty` should toggle the visibility of the f(x) and predict curve whereas `AllOriginalCurvesVisibleProperty` would toggle the visibility of f(x).

The logic for the two boolean expressions would be something like
```js
orignalCurveNode.visible = !predictModeEnabled | (predictModeEnabled && allOriginalCurvesVisible) 
predictCurveNode.visible = predictModeEnabled
```

Another possibility is to have two boolean properties:  `predictModeEnableProperty` and `showOriginalCurveProperty` ;
The visibility of each curve is simply
```js
orignalCurveNode.visible = showOriginalCurve` 
predictCurveNode.visible = predictModeEnabled
```
but every time we toggle `predictModeEnabled`, we assign a new value `showOriginalCurve`
```js
showOriginalCurve =  !predictModeEnabled | (predictModeEnabled && showOriginalCurve) 
```
In other words, we can toggle the state of the 'show f(x)' checkbox by toggling the 'predictModeEnabled' buttons. 

Any thoughts @pixelzoom

**Comment by pixelzoom** [comment]
In OriginalGraphNode.ts, make use of `visibleProperty` to control the visibility of orignalCurveNode and predictCurveNode. Something like this (untested): 

```diff
        curveNodeOptions: {
+         visibleProperty: new DerivedProperty(
+           [ predictModeEnabledProperty, allOriginalCurvesVisibleProperty ],
+          ( predictModeEnabled, allOriginalCurvesVisible ) => !predictModeEnabled | ( predictModeEnabled && allOriginalCurvesVisible )
+         ),
          tandem: providedOptions.tandem.createTandem( 'originalCurveNode' )
        },

...

    const predictCurveNode = new TransformedCurveNode( predictCurve, curveManipulationProperties, this.chartTransform,
      {
+       visibleProperty: predictModeEnabledProperty,
        tandem: providedOptions.tandem.createTandem( 'predictCurveNode' )
      } );
```

**Comment by pixelzoom** [comment]
I should also note that this use of `visibleProperty` is a relatively new pattern for PhET-iO, and is preferred over the old pattern of adding a listener to some Property (or Properties) and calling `node.visible = ...`

**Comment by pixelzoom** [comment]
I should also note that this use of `visibleProperty` is a relatively new (and common!) pattern for PhET-iO. It is preferred over the old pattern of adding a listener to some Property (or Properties) and calling `node.visible = ...`.  And the same pattern is applicable to `enabledProperty`.

So if we have a Node that is only visible when some count is > 0, we might have done this in the past:

```js
const myNode = new MyNode(...);

countProperty.link( count => {
  myNode.visible = ( count > 0 );
} );
```

... but we now prefer to do this:

```js
const myNode = new MyNode( ..., {
  visibleProperty: new DerivedProperty( [ countProperty ], count =>  ( count > 0 ) ),
} );
```

**Comment by veillette** [comment]
I noted your last comment, which led to a lot of small changes to some other visible listeners. (see https://github.com/phetsims/calculus-grapher/commit/308e3119170cc05ee8996dd58b2d1d935a7d148d,  https://github.com/phetsims/calculus-grapher/commit/1ffd444aaede1e34d1ff31eb1bcf7de0f97bbb56,
https://github.com/phetsims/calculus-grapher/commit/6bd6b745d0f8ac3c2215dd9f965532454c7ba1eb).

**Comment by veillette** [comment]
I tried to put most of the logic in a new (model) class to track the color and the selected curve. 
So it is a bit unusual construction , although very useful for the purposes of this simulation. We should try to review on Monday before I get too committed with this approach. 


````typescript
// Copyright 2022, University of Colorado Boulder

/**
 * PredictModeEnabledProperty is Property that track if the predict mode of original graph is enabled
 *
 * @author Martin Veillette
 */

import BooleanProperty from '../../../../axon/js/BooleanProperty.js';
import { PropertyOptions } from '../../../../axon/js/Property.js';
import optionize, { EmptySelfOptions } from '../../../../phet-core/js/optionize.js';
import calculusGrapher from '../../calculusGrapher.js';
import TransformedCurve from './TransformedCurve.js';
import CalculusGrapherColors from '../CalculusGrapherColors.js';
import DerivedProperty from '../../../../axon/js/DerivedProperty.js';
import TReadOnlyProperty from '../../../../axon/js/TReadOnlyProperty.js';
import { Color } from '../../../../scenery/js/imports.js';

type SelfOptions = EmptySelfOptions;

export type SelectedCurveOptions = SelfOptions & PropertyOptions<boolean>;

export default class PredictModeEnabledProperty extends BooleanProperty {

  private readonly originalCurve: TransformedCurve;
  private readonly predictCurve: TransformedCurve;

  public constructor( initialEnabled: boolean,
                      originalCurve: TransformedCurve,
                      predictCurve: TransformedCurve,
                      providedOptions: SelectedCurveOptions ) {

    const options = optionize<SelectedCurveOptions, SelfOptions, PropertyOptions<boolean>>()(
      {}, providedOptions );

    super( initialEnabled, options );

    this.originalCurve = originalCurve;
    this.predictCurve = predictCurve;
  }

  public get curve(): TransformedCurve {
    return this.value ? this.predictCurve : this.originalCurve;
  }

  public get notProperty(): TReadOnlyProperty<boolean> {
    return new DerivedProperty( [ this ], enabled => !enabled );
  }

  public get colorStrokeProperty(): TReadOnlyProperty<Color> {
    return new DerivedProperty( [ this,
        CalculusGrapherColors.predictCurveStrokeProperty,
        CalculusGrapherColors.originalCurveStrokeProperty ],
      ( enabled, predictStroke, originalStroke ) =>
        enabled ? predictStroke : originalStroke );
  }

  // TODO: overkill? or nice to have anyway
  public override reset(): void {
    super.reset();
  }
}
calculusGrapher.register( 'PredictModeEnabledProperty', PredictModeEnabledProperty );
````

**Comment by pixelzoom** [comment]
@veillette and I reviewed https://github.com/phetsims/calculus-grapher/issues/100#issuecomment-1335841086.  The responsibilities in PredictModeEnabledProperty seems a little odd/misplaced for a Property.  So I'm going to take a closer look.

**Comment by pixelzoom** [comment]
> I would also welcome some recommendations for the label inside the f(x) graph updates to say “Predict f(x)” when PredictModeEnabledProperty is set to true.

“Predict f(x)” seems clear to me.  But it's @amanda-phet's call, since she's the designer.

**Comment by pixelzoom** [comment]
Putting responsibilities inside a Property like PredictModeEnabledProperty seems like a bit of a "misplaced responsibilities" design problem.  In the above commits, it was fairly easy to move PredictModeEnabledProperty's responsibilities to other more appropriate places, then replace PredictModeEnabledProperty with a BooleanProperty.   I made these changes as a series of commits, where I eliminated 1 method at a time, so that you could see where the responsibilities were moved to.  Over to @veillette for review, let me know if you have any questions.

**Comment by pixelzoom** [comment]
A minor improvement in the above 2 commits. For DerivedProperty, the parameter names should match the names of the dependencies.

**Comment by veillette** [comment]
In the commit above I created a labelNode in a HBox made of  'Predict' and 'f(x), i.e.  GraphTypeLabelNode( 'original' ). The visibility of the text for 'Predict' is being toggled by the 'predictModeEnabledProperty'.  It seems a simple and clean approach to create two related labels.

**Comment by veillette** [comment]
Assigning to @amanda-phet to review the predict mode feature.

**Comment by amanda-phet** [comment]
I reviewed and everything seems to be behaving as desired! Back to @veillette to close this issue if appropriate.

**Comment by veillette** [comment]
Closing
